src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1125 @@
+        {
+            ChipLogDetail(
+                Zcl, "IsValidAsSupportedLocation %u - LocationName is empty string, FloorNumber, AreaType, LandmarkTag are null",
+                aLocation.locationID);
+            return false;
+        }
+    }
+
+    // If the LandmarkTag field is null, the PositionTag field SHALL be null.
+    if (aLocation.locationInfo.landmarkTag.IsNull() && !aLocation.locationInfo.positionTag.IsNull())
+    {
+        ChipLogDetail(Zcl, "IsValidAsSupportedLocation %u - PositionTag with no LandmarkTag", aLocation.locationID);
+        return false;
+    }
+
+    if (mDelegate->GetNumberOfSupportedMaps() == 0)
+    {
+        // If the SupportedMaps attribute is null, mapid SHALL be null.
+        if (!aLocation.mapID.IsNull())
+        {
+            ChipLogDetail(Zcl, "IsValidSupportedLocation %u - map Id %u is not in empty supported map list", aLocation.locationID,
+                          aLocation.mapID.Value());
+            return false;
+        }
+    }
+    else
+    {
+        // If the SupportedMaps attribute is not null, mapID SHALL be the ID of an entry from the SupportedMaps attribute.
+        if (!IsSupportedMap(aLocation.mapID.Value()))
+        {
+            ChipLogError(Zcl, "IsValidSupportedLocation %u - map Id %u is not in supported map list", aLocation.locationID,
+                         aLocation.mapID.Value());
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Instance::IsUniqueSupportedLocation(const LocationStructureWrapper & aLocation, bool ignoreLocationId)
+{
+    BitMask<LocationStructureWrapper::IsEqualConfig> config;
+
+    if (ignoreLocationId)
+    {
+        config.Set(LocationStructureWrapper::IsEqualConfig::kIgnoreLocationId);
+    }
+
+    // If the SupportedMaps attribute is not null, each entry in this list SHALL have a unique value for the combination of the
+    // MapID and LocationInfo fields. If the SupportedMaps attribute is null, each entry in this list SHALL have a unique value for
+    // the LocationInfo field.
+    if (mDelegate->GetNumberOfSupportedMaps() == 0)
+    {
+        config.Set(LocationStructureWrapper::IsEqualConfig::kIgnoreMapId);
+    }
+
+    uint8_t locationIndex = 0;
+    LocationStructureWrapper entry;
+    while (mDelegate->GetSupportedLocationByIndex(locationIndex++, entry))
+    {
+        if (aLocation.IsEqual(entry, config))
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool Instance::ReportEstimatedEndTimeChange(const DataModel::Nullable<uint32_t> & aEstimatedEndTime)
+{
+    if (mEstimatedEndTime == aEstimatedEndTime)
+    {
+        return false;
+    }
+
+    // The value of this attribute SHALL only be reported in the following cases:
+    // - when it changes from null to any non-zero value
+    if (mEstimatedEndTime.IsNull())
+    {
+        return (aEstimatedEndTime.Value() != 0);
+    }
+
+    if (aEstimatedEndTime.IsNull())
```

Comment:
Again, is it purposeful that changing from not-null-but-not-0 to null does not get reported?

But also, spec says to report when changing from 0 to null, as far as I can tell.  So this does not match spec.

Reply:
Hmm, I'm interpreting "any other value" in "when it changes from 0 to any other value and vice versa" not to include null. @plauric Can you confirm which is the correct interpretation? Is there a need for being more explicit in the spec text here?

@plauric can you explain why changing from not-null-but-not-0 to null does not get reported? You explained this to me but I'm not sure I remember it well.

Reply:
So after conversation with Petru, the "any other value" is meant to include null. This confused me since in the third point the change from null to 0 is explicitly excluded. My thinking was that if you are being this careful to exclude it, the "any other value" must not include null.

To avoid confusion I would propose simplifying the last point to be "when it changes from null". Similarly, the first point could be "when it changes from or to 0."

Reply:
Fixed [here](https://github.com/project-chip/connectedhomeip/pull/33991/commits/af7d1ba17653db04d0fa89b72ade5fb8d0101088).